File: src/main/js/deepdive/functional/monads/Option.js

// Some represents an existing value

// Value might be the type T
// An `Option[T]` is a monad of T or `M[T]`
const Some = (value) => {
    return {
        // T => U - with U as the new type after mapping
        map: mapFunction => Option(mapFunction(value)),

        // T => M[U]
        flatMap: flatMapFunction => flatMapFunction(value),

        get: () => value,

        isPresent: () => true,

        getOrElse: fallbackValue => value
    }
};

// Some represents the nil object
const None = {
    // T => U
    map: mapFunction => None,

    // T => M[U}
    flatMap: flatMapFunction => None,

    get: () => null,

    isPresent: () => false,

    getOrElse: fallbackValue => fallbackValue
};

export const Option = (value) => {
    return value == null ? None : Some(value)
};

File: src/main/kotlin/deepdive/functional/monads/Option.kt
package deepdive.functional.monads

sealed class Option<T> {
    abstract fun <U> map( f: (T) -> U ): Option<U>

    abstract fun <U> flatMap( f: (T) -> Option<U> ): Option<U>

    abstract fun get(): T?

    abstract fun isPresent(): Boolean

    abstract fun getOrElse(fallbackValue: T): T

    companion object {
        fun <T> of(value: T): Option<T> {
            return if(value == null) None() else Some(value)
        }
    }
}


class Some<T> (private val value: T) : Option<T>() {
    override fun <U> map( f: (T) -> U ): Option<U> = Option.of( f(value) )

    override fun <U> flatMap( f: (T) -> Option<U> ): Option<U> = f(value)

    override fun get(): T? = value

    override fun isPresent(): Boolean = true

    override fun getOrElse(fallbackValue: T): T = value
}


class None<T>: Option<T>() {
    override fun <U> map( f: (T) -> U ): Option<U> = this as None<U>

    override fun <U> flatMap( f: (T) -> Option<U> ): Option<U> = this as None<U>

    override fun get(): T? = null

    override fun isPresent(): Boolean = true

    override fun getOrElse(fallbackValue: T): T = fallbackValue
}